### ISSUE TITLE
OpenSSL v1.1.0 hangs ATS

### ISSUE THREAD

**Issue by zwoop** [apache/trafficserver#2183]
I've tested ATS with OpenSSL v1.1.0, and even though it compiles, it does not behave well at all. After a few requests, each ET_NET thread ends up in some sort of loop, consuming all CPUs.

I'll see if I can spend some time again reproducing and identifying this better. If anyone else feels like they want to chip in here and provide more details, please do so :-).

**Comment by unixwitch** [comment]
I think we're hitting this as well, after upgrading from OpenSSL 1.0.1 to 1.1.0.  Thread stack:

```
#0  0x000055e999458d8e in SSLSessionID::operator== (this=0x2add2c252fb0, other=...) at SSLSessionCache.h:78
#1  0x000055e999458be8 in SSLSessionBucket::removeSession (this=0x55e99b393928, id=...) at SSLSessionCache.cc:231
#2  0x000055e9994580ed in SSLSessionCache::removeSession (this=0x55e99b392100, sid=...) at SSLSessionCache.cc:84
#3  0x000055e999435dd1 in ssl_rm_cached_session (ctx=0x55e99b3985f0, sess=0x2add2c000950) at SSLUtils.cc:290
#4  0x00002add23618efa in ?? () from /lib/x86_64-linux-gnu/libssl.so.1.1
#5  0x00002add23619cdf in ?? () from /lib/x86_64-linux-gnu/libssl.so.1.1
#6  0x00002add23614a56 in SSL_free () from /lib/x86_64-linux-gnu/libssl.so.1.1
#7  0x000055e99942fe12 in SSLNetVConnection::free (this=0x2add4c01af20, t=0x2add2597f010) at SSLNetVConnection.cc:888
#8  0x000055e9994492aa in close_UnixNetVConnection (vc=0x2add4c01af20, t=0x2add2597f010) at UnixNetVConnection.cc:134
#9  0x000055e999449481 in read_signal_and_update (event=104, vc=0x2add4c01af20) at UnixNetVConnection.cc:164
#10 0x000055e9994496a3 in read_signal_done (event=104, nh=0x2add25982e60, vc=0x2add4c01af20) at UnixNetVConnection.cc:206
#11 0x000055e99944c3d8 in UnixNetVConnection::readSignalDone (this=0x2add4c01af20, event=104, nh=0x2add25982e60)
    at UnixNetVConnection.cc:1088
#12 0x000055e99942e994 in SSLNetVConnection::net_read_io (this=0x2add4c01af20, nh=0x2add25982e60, lthread=0x2add2597f010)
    at SSLNetVConnection.cc:628
#13 0x000055e999440c9f in NetHandler::mainNetEvent (this=0x2add25982e60, event=5, e=0x55e99b2a4d60) at UnixNet.cc:499
#14 0x000055e9991c2ba2 in Continuation::handleEvent (this=0x2add25982e60, event=5, data=0x55e99b2a4d60)
    at /usr/src/trafficserver-7.1.x/iocore/eventsystem/I_Continuation.h:153
#15 0x000055e99946c1f1 in EThread::process_event (this=0x2add2597f010, e=0x55e99b2a4d60, calling_code=5) at UnixEThread.cc:143
#16 0x000055e99946c6f7 in EThread::execute (this=0x2add2597f010) at UnixEThread.cc:270
#17 0x000055e99946b8a5 in spawn_thread_internal (a=0x55e99b280550) at Thread.cc:84
#18 0x00002add23ce9494 in start_thread (arg=0x2add25b80700) at pthread_create.c:333
#19 0x00002add24a88aff in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:97
```

The hangup is here: https://github.com/apache/trafficserver/blob/7.1.x/iocore/net/SSLSessionCache.cc#L230; if `node->session_id != id` then it will busy-loop forever.

Disabling the session cache seems to fix it.

**Comment by shinrich** [comment]
Yes, that looks like a very bad loop.

**Comment by zwoop** [comment]
I'm still seeing this after #2183

**Comment by zwoop** [comment]
This is fixed, and in 7.1.0 release.
